server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -125,7 +127,27 @@ public TransportBulkAction(Settings settings, ThreadPool threadPool, TransportSe
 
     @Override
     protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
-        if (bulkRequest.hasIndexRequestsWithPipelines()) {
+        boolean hasIndexRequestsWithPipelines = false;
+        ImmutableOpenMap<String, IndexMetaData> indicesMetaData = clusterService.state().getMetaData().indices();
+        for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
+            if (actionRequest instanceof IndexRequest) {
+                IndexRequest indexRequest = (IndexRequest) actionRequest;
+                String pipeline = indexRequest.getPipeline();
+                if (pipeline == null) {
+                    IndexMetaData indexMetaData = indicesMetaData.get(indexRequest.index());
+                    if (indexMetaData != null) {
+                        String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexMetaData.getSettings());
+                        if (!defaultPipeline.isEmpty()) {
+                            indexRequest.setPipeline(defaultPipeline);
+                            hasIndexRequestsWithPipelines = true;
+                        }
+                    }
+                } else if (!pipeline.isEmpty()) {
```

Comment:
We prefer `== false` instead of `!` as the latter is susceptible to being missed when reading through code.